### PR TITLE
Avoid needless allocation in read-bytevector!

### PR DESCRIPTION
**Pull request by wasamasa**
This change switches the implementation strategy to basing read-bytevector on top of read-bytevector! rather than the other way around. Bonus: It's a bit less code than previously and is less surprising to the reader.

### CODE REVIEW

**Review by ashinn — CHANGES_REQUESTED** [review]
Thanks!  I'm not sure why I did it this way - it should really be done in the FFI but your approach is better until I get around to that.

File: lib/scheme/extras.scm [hunk 1]
```diff
@@ -136,35 +136,28 @@
...
-        (let lp ((i 0))
-          (if (>= i n)
-              res
-              (let ((x (read-u8 in)))
-                (cond ((eof-object? x)
-                       (if (zero? i) x (subbytes res 0 i)))
-                      (else
-                       (bytevector-u8-set! res i x)
-                       (lp (+ i 1))))))))))
+        (read-bytevector! res in)
```
[the remark this answers is not in the document]
Comment (wasamasa) [reply lib/scheme/extras.scm]: Good point, fixed

File: lib/scheme/extras.scm [hunk 2]
```diff
@@ -135,36 +135,31 @@
   (if (zero? n)
       #u8()
       (let ((in (if (pair? o) (car o) (current-input-port)))
```
Comment (ashinn) [inline lib/scheme/extras.scm:137]: `let*`

**Review by ashinn — APPROVED** [review]
Thanks!

File: lib/scheme/extras.scm [hunk 3]
```diff
@@ -134,37 +134,32 @@
...
-                       (if (zero? i) x (subbytes res 0 i)))
-                      (else
-                       (bytevector-u8-set! res i x)
-                       (lp (+ i 1))))))))))
+      (let* ((in (if (pair? o) (car o) (current-input-port)))
+             (vec (make-bytevector n))
+             (res (read-bytevector! vec in)))
+        (cond ((eof-object? res) res)
+              ((< res n) (subbytes vec 0 i))
+              (else res)))))
```
Comment (ashinn) [inline lib/scheme/extras.scm:142]:
```suggestion
              (else vec)))))
```

File: lib/scheme/extras.scm [hunk 4]
```diff
@@ -134,37 +134,32 @@
...
-                (cond ((eof-object? x)
-                       (if (zero? i) x (subbytes res 0 i)))
-                      (else
-                       (bytevector-u8-set! res i x)
-                       (lp (+ i 1))))))))))
+      (let* ((in (if (pair? o) (car o) (current-input-port)))
+             (vec (make-bytevector n))
+             (res (read-bytevector! vec in)))
+        (cond ((eof-object? res) res)
+              ((< res n) (subbytes vec 0 i))
```
Comment (ashinn) [inline lib/scheme/extras.scm:141]:
```suggestion
              ((< res n) (subbytes vec 0 res))
```

**Review by ashinn — CHANGES_REQUESTED** [review]

**Pull request merged**